### PR TITLE
retry downloading core and sync binaries

### PR DESCRIPTION
**Pull request by larkost**

### CODE REVIEW

**Comment by jpsim** [comment]
Is this ready for review and can you please share examples of CI failures that this should fix?

**Comment by larkost** [comment]
I am trying to puzzle out what is happening here, since the `tvos Release 8.0` is showing the exact error that this was trying to avoid... in a way that should no longer be possible (the message has changed). So... not yet, and the example: https://ci.realm.io/job/objc_pr/configuration=Release,target=tvos,xcode_version=8.0/4981/console

**Comment by jpsim** [comment]
Alright. I'll mark this as `S:In Progress` then to reflect its current state.

I'd appreciate if you could read our [GitHub Issues Usage](https://github.com/realm/realm-wiki/wiki/Realm-GitHub-Issues-Usage) wiki article which outlines how we use labels as part of our engineering management process.

**Comment by larkost** [comment]
This time it tested the correct version, so things look ready to go.

**Comment by jpsim** [comment]
> This time it tested the correct version, so things look ready to go.

As indicated in [GitHub Issues Usage](https://github.com/realm/realm-wiki/wiki/Realm-GitHub-Issues-Usage) wiki article, this is when you'd change the label from `S:In Progress` to `S:Review`. I've done that this time and am reviewing this now.

File: build.sh [hunk 1]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
-    if [ ! -f "${CORE_TAR}" ]; then
-        local CORE_URL="https://static.realm.io/downloads/core/realm-core-${REALM_CORE_VERSION}.tar.xz"
-        set +e # temporarily disable immediate exit
-        local ERROR # sweeps the exit code unless declared separately
-        ERROR=$(curl --fail --silent --show-error --location "$CORE_URL" --output "${CORE_TMP_TAR}" 2>&1 >/dev/null)
-        if [[ $? -ne 0 ]]; then
-            echo "Downloading core failed:\n${ERROR}"
-            exit 1
+download_common() {
+    local download_type=$1 tries_left=3 version core_url error temp_dir temp_path tar_path
```
Comment (jpsim) [inline build.sh]: `core_url` should be `url`

File: build.sh [hunk 2]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
+        echo "Unknown dowload_type: $download_type"
+        exit 1
+    fi
+    
+    echo "Downloading dependency: ${download_type} ${version}"
+    
+    if [ -z "$TMPDIR" ]; then
+        TMPDIR='/tmp'
+    fi
+    temp_dir=$(dirname "$TMPDIR/waste")/core_bin
```
Comment (jpsim) [inline build.sh]:
```bash
temp_dir="$(dirname "$TMPDIR/waste")/${download_type}_bin"
```

File: build.sh [hunk 3]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
+    mkdir -p "$temp_dir"
+    temp_path="${temp_dir}/${download_type}-${version}.tar.xz.tmp"
```
Comment (jpsim) [inline build.sh]: Flip these?

```bash
tar_path="${temp_dir}/${download_type}-${version}.tar.xz"
temp_path="${tar_path}.tmp"
```

File: build.sh [hunk 4]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
+    tar_path="${temp_dir}/${download_type}-${version}.tar.xz"
+        
+    while [ 0 -lt $tries_left ] && [ ! -f "$tar_path" ]; do
+        if ! error=$(curl --fail --silent --show-error --location "$url" --output "$temp_path" 2>&1); then
+            tries_left=$[$tries_left-1]
```
Comment (jpsim) [inline build.sh]: according to [shellcheck](https://www.shellcheck.net), this should be `tries_left=$($tries_left-1)`

File: build.sh [hunk 5]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
+        else
+            mv "$temp_path" "$tar_path"
         fi
-        set -e # re-enable flag
-        mv "${CORE_TMP_TAR}" "${CORE_TAR}"
+    done
+    
+    if [ ! -f $tar_path ]; then
```
Comment (jpsim) [inline build.sh]: should be quoted: `"$tar_path"`

File: build.sh [hunk 6]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
+        printf "Downloading ${download_type} failed:\n\t$url\n\t$error\n"
```
Comment (jpsim) [inline build.sh]: according to [shellcheck](https://www.shellcheck.net), format strings are preferred with `printf`:

```bash
printf "Downloading %s failed:\n\t%s\n\t%s\n" "$download_type" "$url" "$error"
```

File: build.sh [hunk 7]
```diff
@@ -321,66 +321,61 @@ download_object_server() {
...
     fi
-
+    
     (
-        cd "${TMP_DIR}"
-        rm -rf core
-        tar xf "${CORE_TAR}" --xz
-        mv core core-${REALM_CORE_VERSION}
+        cd "$temp_dir"
+        rm -rf $download_type
```
Comment (jpsim) [inline build.sh]: fine today, but might as well stay safe in the future by quoting this: `rm -rf "$download_type"`

File: build.sh (hunk 6 above)
Comment (larkost) [inline build.sh]: That is both harder to read in this case, and gets us into potential problems between GNU and BSD versions of bash (the former requires commas, the latter confusingly prints the commas). In theory using the format string would avoid some injection-style stacks, but I don't think we are worried about those here.

File: build.sh [hunk 8]
```diff
@@ -340,27 +340,27 @@ download_common() {
...
         if ! error=$(curl --fail --silent --show-error --location "$url" --output "$temp_path" 2>&1); then
-            tries_left=$[$tries_left-1]
+            tries_left=$($tries_left-1)
         else
             mv "$temp_path" "$tar_path"
         fi
     done
     
-    if [ ! -f $tar_path ]; then
+    if [ ! -f '$tar_path" ]; then
```
Comment (jpsim) [inline build.sh]: can't mix single quotes with double quotes

**Pull request merged**